### PR TITLE
disk backed dataset, k8s running, and splitting data

### PR DESCRIPTION
**Pull request by blester125**
This PR does the following:

 * Adds some logging to the data reading (vocab generation and featurization) so we know where things go wrong if there is a crash
 * Fixes a bug where utterances weren't being `.split()`ed when loading the features
 * Fixes a bug where the rank would cause trying to use a gpu you don't have in an environment like k8s where you only ever get gpu `0`
 * Adds a disk backed reader and pytorch dataset object. This lets us avoid crashes I was seeing then trying to read the whole dataset into memory in the `load_features` method
 * Adds a jagged tensor reader. Tensors are stored in a list so they don't have to be padded. there is then a `pad_batch` function used as the `collate_fn` for the `DataLoader` this lets us keep the whole dataset in memory, padded tensors take up too much space because the average utt only takes about 1/4 of the mxlen

**Currently I am not able to get it to train with the all loss, I am looking into this**

### CODE REVIEW

File: api-examples/pretrain-paired.py [hunk 1]
```diff
@@ -290,6 +296,145 @@ def load(self, filename, vocabs):
         y_tensor = torch.tensor(features['y'], dtype=torch.long)
         return TensorDataset(x_tensor, y_tensor)
 
+class TensorWordJaggedDatasetReader(TensorWordDatasetReader):
```
Comment (dpressel) [inline api-examples/pretrain-paired.py]: lets not do this unless there is a really good reason
